src/server/bg_services/tier_ttl_worker.js

```diff
@@ -0,0 +1,123 @@
+ * worker relies on `evict_block` RPC which implies that the associated blocks
+ * stores must implement the `evict_block` RPC.
+ */
+class TieringTTLWorker {
+    constructor({ name, client }) {
+        this.name = name;
+        this.client = client;
+        this.initialized = false;
+        this.last_run = 'force';
+    }
+
+    _can_run() {
+        if (!system_store.is_finished_initial_load) {
+            dbg.log0('TieringTTLWorker: system_store did not finish initial load');
+            return false;
+        }
+
+        const system = system_store.data.systems[0];
+        if (!system || system_utils.system_in_maintenance(system._id)) return false;
+
+        return true;
+    }
+
+    async run_batch() {
+        if (!this._can_run()) return;
+
+        dbg.log0('TieringTTLWorker: start running');
+        const candidate_buckets = this._get_candidate_buckets();
+        if (!candidate_buckets || !candidate_buckets.length) {
+            dbg.log0('no buckets with more than one tier. nothing to do');
+            this.last_run = 'force';
+            return config.TIERING_TTL_WORKER_BATCH_DELAY;
+        }
+
+        const wait = await this._ttl_move_chunks(candidate_buckets);
+        console.log(`TieringTTLWorker: will wait ${wait} ms till next run`);
+        return wait;
+    }
+
+    _get_candidate_buckets() {
+        return system_store.data.buckets.filter(bucket =>
+            !bucket.deleting && (
+                // only include buckets that have 2 or more tiers
+                (bucket.tiering && bucket.tiering.tiers.length > 1)
+            ));
+    }
+
+    async _ttl_move_chunks(buckets) {
+        const chunks_to_move = config.TIERING_TTL_WORKER_BATCH_SIZE;
+
+        for (const bucket of buckets) {
+            await node_allocator.refresh_tiering_alloc(bucket.tiering, this.last_run);
+            const tiering_status = node_allocator.get_tiering_status(bucket.tiering);
+            const previous_tier = mapper.select_tier_for_write(bucket.tiering, tiering_status);
+            const next_tier_order = this.find_tier_order_in_tiering(bucket, previous_tier) + 1;
+            const next_tier = mapper.select_tier_for_write(bucket.tiering, tiering_status, next_tier_order);
+
+            // no point in calling build_chunks when there is no next_tier
+            if (!next_tier) continue;
+
+            const next_tier_id = next_tier._id;
+
+            const chunk_ids = await MDStore.instance().find_oldest_tier_chunk_ids({
+                tier: previous_tier._id,
+                limit: chunks_to_move,
+                sort_direction: 1,
+                max_tier_lru: moment().subtract(config.TIERING_TTL_MS, 'ms').toDate(),
+            });
+            if (!chunk_ids.length) continue;
+
+            await this._build_chunks(
+                chunk_ids,
+                next_tier_id,
+                false,
+                chunk_ids.map(() => previous_tier._id)
```

Comment:
if we always have just a single tier here, then do we really want to send an array?
we could just send the previous_tier_id just like we do for the next_tier_id.
I think maybe this array of current_tiers feels a bit too much.

Reply:
I made this an array under the assumption that the callers of the scrubber.build_chunks do not guarantee that the current tier of all the chunks is same.

If the above assumption is not true then maybe we can switch to a single value parameter.